core/src/main/scala/loadbalancers/Aws.scala

```diff
@@ -68,40 +65,179 @@ final class Aws(cfg: Infrastructure.Aws) extends (LoadbalancerOp ~> IO) {
+    log.debug(s"calling aws client to launch $name")
     val size = asgSize(p)
-    createELB(name, lb.routes.map(_.port), cfg.lbScheme) <* createASG(name, ns, size)
+
+
+    for {
+      loadbalancer <- createNLB(name, cfg.lbScheme)
+
+      targetGroups <- createTargetGroups(name, lb.routes.map(_.port))
+      _ <- createListeners(loadbalancer, targetGroups)
+
+      _ <- createASG(name, ns, size)
+      _ <- attachASGToTargetGroups(name, targetGroups)
+    } yield loadbalancer.getDNSName
   }
 
+  // The target group has this restriction where
+  // the name of the target group can't have "double dashes". I want the loadbalancer name
+  // to match the target group name, so the loadbalancer won't have the double dashes either.
   def loadbalancerName(name: String, v: MajorVersion, hash: String) =
-    s"$name--${v.minVersion.toExternalString}--${hash}"
+    s"$name-${v.minVersion.toExternalString}-${hash}"
+
+  def getLoadBalancerArn(name: String): IO[String] = {
+    val describeLoadBalancerRequest = new DescribeLoadBalancersRequest()
+      .withNames(name)
+    IO(
+      cfg.nlb.describeLoadBalancers(describeLoadBalancerRequest)
+    ).map(resp => resp.getLoadBalancers.asScala.toList)
+      .map(loadbalancerList => loadbalancerList.head.getLoadBalancerArn).recoverWith{
+      case e: Exception =>
+        log.error(s"aws call to get loadbalancer ARN ${describeLoadBalancerRequest.getNames.asScala.head} failed.")
+        IO.raiseError(FailedLoadbalancerDeploy(s"could not find loadbalancer and therefore could not delete loadbalancer", e.getMessage))
+    }
+  }
 
-  def deleteELB(name: String): IO[Unit] =
-    IO {
-      val req = new DeleteLoadBalancerRequest(name)
-      cfg.elb.deleteLoadBalancer(req)
+  def deleteNLB(lbarn: String): IO[Unit] = {
+    IO{
+      val req: DeleteLoadBalancerRequest = new DeleteLoadBalancerRequest()
+        .withLoadBalancerArn(lbarn)
+
+      cfg.nlb.deleteLoadBalancer(req)
       ()
     }
+  }
 
-  def createELB(name: String, p: Vector[Port], scheme: ElbScheme): IO[DNSName] =
+  // https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/elasticloadbalancingv2/model/CreateLoadBalancerResult.html
+  // has no method getDNS().
+  def createNLB(name: String, scheme: NlbScheme): IO[LoadBalancer] = {
     IO {
-      val listeners = p.map(p => new Listener(TCP,p.port,p.port)).toList
-      val req = new CreateLoadBalancerRequest(name)
-        .withListeners(listeners.asJava)
+      val req = new CreateLoadBalancerRequest()
+        .withName(name)
+        .withType(LoadBalancerTypeEnum.Network)
         .withScheme(scheme.toString)
-        .withSecurityGroups(cfg.elbSecurityGroupNames.toList.asJava)
         .withSubnets(cfg.availabilityZones.map(_.publicSubnet).asJava)
 
-      val resp = cfg.elb.createLoadBalancer(req)
+      val resp = cfg.nlb.createLoadBalancer(req)
+      val loadbalancers: List[LoadBalancer] = resp.getLoadBalancers.asScala.toList
+
+      loadbalancers.head
```

Comment:
This still makes me uncomfortable. @adelbertc got any suggestions here to avoid using `head`? We could do some gymnastics wiith `headOption` and monad transformers, but im not sure its worth it.

Reply:
This would do the trick, but unsure how much better it is.
```
Alternative[Option]
  .guard(loadbalancers.size === 1)
  .flatMap(loadbalancers.headOption)
  .fold(IO.raiseError(...))(_.pure[IO])
```